README.md
# Pact

[![Build Status](https://secure.travis-ci.org/reid/pact.png)](http://travis-ci.org/reid/pact)

A collection of [Vows][] macros for easy HTTP server testing.

Works with [Node.js](http://nodejs.org/) v0.2.5 and later.

## Documentation

<http://reid.github.com/pact/>

## Installation

    npm i pact

## Example

    // Pact works with http.Server instances.
    // This includes express.Server, connect.Server, etc.
    // This method returns a new express.Server:
    var createServer = require("../lib/app").createServer;

    require("vows").describe("HTTP Server").addBatch({
        "A server in development" : {
            // Start a server for testing with httpify
            // Give it a new http.Server
            topic : httpify(createServer()),
            "when /foo is requested" : {
                topic : request(), // knows the URL from context name
                "should fail" : code(400) // check status code
            }
            "when /foo?bar=baz is requested" : {
                topic : request(),
                "should succeed" : code(200),
                "should return response time header" : function (topic) {
                    // header names are lowercased for easy testing
                    assert.include(topic.headers, "x-response-time");
                },
                "should be correct size" : function (topic) {
                    // response is available as topic.body
                    assert.equal(topic.body.length, 11);
                }
            },
            "when making a bogus request" : {
                // you can always specify your own URL
                // POST requests work as well
                topic : request({
                    url : "/bogus",
                    method : "POST",
                    body : "quux=0"
                }),
                "should fail" : code(404)
            }
        },
        "A server in production" : {
            topic : function () {
                // Example: wrap httpify for testing
                // with a new environment
                var oldEnv = process.env.NODE_ENV;
                process.env.NODE_ENV = "production";
                var server = app.createServer();
                httpify(server).apply(this);
                process.env.NODE_ENV = oldEnv;
            },
            "when / is requested" : {
                topic : request(),
                "should fail" : code(404)
            }
        }
    }).export(module);

## Run self-tests

Requires [Vows][].

    make test

## About

Authored by [Reid Burke](http://github.com/reid), copyright Yahoo! Inc., and provided under the BSD license. See LICENSE file.

## History

Pact is used at Yahoo! for testing Node.js servers. It's based on the embedded Vows macros from the YUI Labs [Yeti](http://github.com/yui/yeti) project.

## Bugs

Submit bugs and pull requests to [Pact on GitHub](http://github.com/reid/pact).

[Vows]: http://vowsjs.org/


HISTORY.md
# Pact History

## Version 0.1.3 (2013-10-02)

* Add support for node v0.8 and v0.10 by using `http.request` if `http.createClient` is deprecated.

## Version 0.1.2 (2011-03-29)

* Parse JSON responses even when the content-type includes a charset. (Ryan Grove)

## Version 0.1.1 (2011-03-29)

* You can now specify HTTP headers in request().
* Only follow relative 302 redirects.
* Linted.

## Version 0.1.0 (2011-02-08)

* First public release.


test.js
/*!
 * Pact
 * Copyright 2011 Yahoo! Inc.
 * Licensed under the BSD license.
 */

var vows = require('vows');
var assert = require('assert');

var pact = require('./index');

var http = require('http');

function mockServer(cb, port) {
    return function() {
        pact.httpify(http.createServer(function(req, res) {
            cb(req, res, port);
        }), port).apply(this);
    };
}

var headers = { 'Content-Type' : 'text/plain' };

vows.describe('Pact').addBatch({
    'A simple server' : {
        topic: mockServer(function(req, res) {
            res.writeHead(200, headers);
            res.end('ok!');
        }),
        'when / is requested' : {
            topic: pact.request(),
            'should succeed' : function(topic) {
                assert.strictEqual(topic.status, 200);
            },
            'contains correct header' : function(topic) {
                assert.include(topic.headers, 'content-type');
                assert.strictEqual(topic.headers['content-type'],
                    headers['Content-Type']);
            },
            'contains valid data' : function(topic) {
                assert.strictEqual(topic.body, 'ok!');
            }
        },
    },
    'A JSON API' : {
        topic: mockServer(function(req, res) {
            if (req.url === '/') {
                res.writeHead(200, {'Content-Type': 'application/json'});
            } else if (req.url == '/with-charset') {
                res.writeHead(200, {'Content-Type': 'application/json;charset=utf-8'});
            }
            res.end('{"foo": "bar"}');
        }),
        'when / is requested' : {
            topic: pact.request(),
            'contains a parsed JSON response object' : function(topic) {
                assert.isObject(topic.body);
                assert.strictEqual(topic.body.foo, 'bar');
            }
        },
        'when /with-charset is requested' : {
            topic: pact.request(),
            'contains a parsed JSON response object' : function(topic) {
                assert.isObject(topic.body);
                assert.strictEqual(topic.body.foo, 'bar');
            }
        }
    },
    'A 302 redirecting server with a relative path' : {
        topic: mockServer(function(req, res) {
            if (req.url === '/ok') {
                res.writeHead(200, headers);
                res.end('tango');
            } else {
                res.writeHead(302, {
                    'Location' : '/ok'
                });
                res.end();
            }
        }),
        'when / is requested' : {
            topic: pact.request(),
            'should be 200 instead of 302' : function(topic) {
                assert.strictEqual(topic.status, 200);
            },
            'contains valid data' : function(topic) {
                assert.strictEqual(topic.body, 'tango');
            }
        }
    },
    'A 302 redirecting server with an absolute path' : {
        topic: mockServer(function(req, res, port) {
            if (req.url === '/ok') {
                res.writeHead(200, headers);
                res.end('foxtrot');
            } else if (req.url == '/') {
                res.writeHead(302, {
                    'Location' : 'http://127.0.0.1:' + port + '/ok'
                });
                res.end();
            } else {
                res.writeHead(404, headers);
                res.end('Invalid path: ' + req.url);
            }
        }, 8091),
        'when / is requested' : {
            topic: pact.request(),
            'should be 302' : function(topic) {
                assert.strictEqual(topic.status, 302);
            },
            'contains Location with empty body' : function(topic) {
                assert.include(topic.headers, 'location');
                assert.strictEqual(topic.body, '');
            }
        }
    }
}).export(module);


lib/http.js
/*!
 * Pact HTTP
 * Copyright 2011 Yahoo! Inc.
 * Licensed under the BSD license.
 */

var http = require("http");
var EventEmitter = require("events").EventEmitter;

// High level response `EventEmitter` for an HTTP request.
// Fires `response` and `error` events only.
function responseEventForRequest (req) {
    var event = new EventEmitter();
    req.on("response", function (response) {
        response.setEncoding("utf8");
        var data = "";
        response.on("data", function (chunk) {
            data += chunk;
        });
        response.on("end", function () {
            if (/^application\/json\s*(?:;|$)/.test(response.headers["content-type"])) {
                data = JSON.parse(data);
            }
            event.emit("response", response, data);
        });
    });
    if (req.connection) { // node v0.2.x
        req.connection.on("error", function (err) {
            event.emit("error", err);
        });
    }
    req.end();
    return event;
}

// An elegant wrapper to `http.createClient` and more.
// Sets appropiate headers, encodes body objects as JSON.
// Returns a `responseEventForRequest`.
function httpRequest (d) {

    d.method = d.method || "GET";
    d.headers = d.headers || {};
    d.headers["host"] = d.host;

    if ("object" === typeof d.body) {
        d.body = JSON.stringify(d.body);
        d.headers["content-type"] = "application/json";
    }

    if (d.body) {
        d.headers["content-length"] = d.body.length;
    }

    var req;
    // Node compat shim, http.createClient was deprecated
    if (http.request) {
        req = http.request({
            host:    d.host,
            port:    d.port || (d.secure ? 443 : 80),
            method:  d.method,
            path:    d.path,
            headers: d.headers
        });
    } else {
        req = http.createClient(
            d.port || (d.secure ? 443 : 80),
            d.host,
            d.secure
        ).request(d.method, d.path, d.headers);
    }

    if (d.body) req.write(d.body);

    return responseEventForRequest(req);
}

// Provide as `http.request`.
exports.request = httpRequest;


index.js
/*!
 * Pact
 * Copyright 2011 Yahoo! Inc.
 * Licensed under the BSD license.
 */

/**
 * Dependencies.
 */
var assert = require('assert');

var http = require('./lib/http');
var STATUS_CODES = require('http').STATUS_CODES;

/**
 * A starting server port number.
 */
var _port = 19100;

/**
 * Factory for server port numbers.
 *
 * @return {Number} port A new port number.
 */
function getPort() {
    return ++_port;
}

/**
 * Return a function that makes an HTTP request
 * according to what's specified in the provided `req`
 * object.
 *
 * Relative 302 redirects will be followed.
 *
 * You can specify:
 *
 * - url: The path to request. Defaults to a path as the second word of
 *   a context.
 * - method: Defaults to GET.
 * - data: Request body for POST.
 * - headers: HTTP request headers object.
 *
 * These are all optional.
 *
 * Instead of specifying `url` in `req`, you can make it the second word of
 * your context. This lets you omit `req` completely, for example:
 *
 *     "when /foo/bar/baz is requested" : {
 *          topic : request(),
 *          "should succeed" : code(200)
 *     }
 *
 * Your test functions will recieve an object containing:
 *
 *  - body: The response body, which is an object if the response was
 *    application/json.
 *  - status: HTTP status code.
 *  - headers: HTTP headers as an object, with headers in lowercase.
 *
 * A topic function.
 *
 * For an example test function, {@see code}.
 *
 * @param {Object} req Request object.
 * @return {Function} Topic function that makes the request.
 */
function request(req) {
    var path, options = {
        host: 'localhost',
        method: 'GET'
    };
    if (req) {
        if ('url' in req) path = options.path = req.url;
        if ('data' in req) options.body = req.data;
        if ('method' in req) options.method = req.method;
        if ('headers' in req) options.headers = req.headers;
    }
    return function(lastTopic) {
        var vow = this;
        // try to find port number
        var port = Array.prototype.slice.call(arguments, -1)[0];
        if (!isNaN(port))
            options.port = port;
        else throw new Error('Unable to determine port from topic.');
        if ('function' === typeof path)
            options.path = path(lastTopic);
        else if (!path)
            options.path = vow.context.name.split(/ +/)[1];
        http.request(
            options
        ).on('response', function X(res, results) {
            var err = null;
            if (res.statusCode === 302) { // handle redirects
                if (options._302) {
                    err = 'Redirect loop';
                } else {
                    var location = res.headers.location;
                    if (location.indexOf('/') === 0) {
                        // relative path, don't handle absolute
                        options.path = location;
                        options._302 = true;
                        return http.request(options).on('response', X);
                    }
                }
            }
            vow.callback(err, {
                body: results,
                options: options,
                status: res.statusCode,
                headers: res.headers
            });
        });
    }
}

/**
 * Return a function that starts a `http.Server` listening on 127.0.0.1.
 *
 * Your test functions will recieve the port the server is listening on.
 *
 * A topic function.
 *
 * @param {Object} server `http.Server` instance.
 * @param {Number} port Optional. Port to listen on.
 * @return {Function} Topic function that starts the server.
 */
function httpify(server, port) {
    port = port || getPort();
    return function() {
        var vows = this;
        server.listen(port, '127.0.0.1', function(err) {
            vows.callback(err, port);
        });
    };
}

/**
 * Return a function that asserts that `status` matches httpify's status.
 *
 * A test function.
 *
 * @see httpify
 * @param {Number} status Expected HTTP status code.
 * @return {Function} Test function that checks the status.
 */
function code(status) {
    return function(lastTopic) {
        assert.strictEqual(lastTopic.status, status,
            'Expected ' + status +
                  ' ' + STATUS_CODES[status] +
                  ', received: ' + lastTopic.status +
                  ' ' + STATUS_CODES[lastTopic.status] +
                  '.\n' + lastTopic.body
        );
    };
}

/**
 * Export these functions.
 */
module.exports = {
    code: code,
    request: request,
    httpify: httpify
};
